File: bindings/ruby/kryptic.rb
# frozen_string_literal: true

# Kryptic Ruby Client
#
# Communicates with a running Kryptic server (python -m kryptic serve).
# Zero external dependencies — uses Ruby's built-in Net::HTTP.
#
# Requirements: Ruby 2.7+
#
# Usage:
#   require_relative 'kryptic'
#   k = Kryptic::Client.new
#   s = k.session
#   s.goto('https://example.com')
#   puts s.title
#   s.close

require 'net/http'
require 'json'
require 'base64'
require 'uri'

module Kryptic
  class Error < StandardError; end

  class Client
    def initialize(host: '127.0.0.1', port: 7890)
      @host = host
      @port = port
      @http = Net::HTTP.new(host, port)
      @http.read_timeout = 60
      @http.open_timeout = 10
    end

    # ── Low-level helpers ──────────────────────────────────────────────────────

    def request(method, path, body = nil)
      uri = URI::HTTP.build(host: @host, port: @port, path: path)

      req_class = case method.upcase
                  when 'GET'    then Net::HTTP::Get
                  when 'POST'   then Net::HTTP::Post
                  when 'DELETE' then Net::HTTP::Delete
                  else raise ArgumentError, "Unsupported method: #{method}"
                  end

      req = req_class.new(uri)
      req['Content-Type'] = 'application/json'
      req.body = body ? body.to_json : '{}'

      resp = @http.request(req)
      data = JSON.parse(resp.body)

      raise Error, data['error'] || 'Unknown error' unless data['ok']

      data
    end

    def get(path)
      uri = URI::HTTP.build(host: @host, port: @port, path: path)
      req = Net::HTTP::Get.new(uri)
      req['Content-Type'] = 'application/json'
      resp = @http.request(req)
      JSON.parse(resp.body)
    end

    # ── API ────────────────────────────────────────────────────────────────────

    def health
      get('/health')
    end

    def session
      data = request('POST', '/sessions')
      Session.new(self, data['session_id'])
    end

    def http_get(url, headers: {})
      request('POST', '/http/get', { url: url, headers: headers })
    end

    def http_post(url, json: {}, headers: {})
      request('POST', '/http/post', { url: url, json: json, headers: headers })
    end

    def http_batch(urls)
      data = request('POST', '/http/batch', { urls: urls })
      data['results'] || []
    end
  end

  class Session
    attr_reader :id

    def initialize(client, id)
      @client = client
      @id     = id
    end

    def post(action, body = {})
      @client.request('POST', "/sessions/#{@id}/#{action}", body)
    end

    def get_action(action)
      @client.get("/sessions/#{@id}/#{action}")
    end

    def goto(url, wait_until: 'domcontentloaded')
      post('goto', { url: url, wait_until: wait_until })
    end

    def title
      get_action('title')['title'] || ''
    end

    def html
      get_action('html')['html'] || ''
    end

    def url
      get_action('url')['url'] || ''
    end

    def text(selector)
      post('text', { selector: selector })['text'] || ''
    end

    def click(selector)
      post('click', { selector: selector })
    end

    def fill(selector, value)
      post('fill', { selector: selector, value: value })
    end

    def evaluate(js)
      post('evaluate', { js: js })['result']
    end

    def find(selector)
      post('find', { selector: selector })
    end

    def screenshot(full_page: false)
      data = post('screenshot', { full_page: full_page })
      Base64.decode64(data['data'] || '')
    end

    def block_resources(types: %w[image stylesheet font media])
      post('block', { resource_types: types })
    end

    def wait_for(selector, state: 'visible')
      post('wait_for', { selector: selector, state: state })
    end

    def close
      @client.request('DELETE', "/sessions/#{@id}")
    end
  end
end

# ── Example ────────────────────────────────────────────────────────────────────

if __FILE__ == $PROGRAM_NAME
  k = Kryptic::Client.new
  puts "Health: #{k.health}"

  s = k.session
  s.block_resources
  s.goto('https://example.com')
  puts "Title: #{s.title}"
  puts "H1:    #{s.text('h1')}"
  s.close

  resp = k.http_get('https://httpbin.org/get')
  puts "HTTP status: #{resp['status']}"

  batch = k.http_batch(%w[https://example.com https://example.org https://iana.org])
  batch.each { |r| puts "  #{r['status']}  #{r['url']}" }
end

File: bindings/php/kryptic.php
<?php
/**
 * Kryptic PHP Client
 *
 * Communicates with a running Kryptic server (python -m kryptic serve).
 * Zero external dependencies — uses PHP's built-in cURL extension.
 *
 * Requirements: PHP 7.4+, ext-curl, ext-json
 *
 * Usage:
 *   $k = new KrypticClient();
 *   $session = $k->session();
 *   $session->goto('https://example.com');
 *   echo $session->title();
 *   $session->close();
 */

class KrypticException extends RuntimeException {}

class KrypticClient
{
    private string $base;

    public function __construct(string $host = '127.0.0.1', int $port = 7890)
    {
        $this->base = "http://{$host}:{$port}";
    }

    // ── Low-level request helpers ────────────────────────────────────────────

    public function request(string $method, string $path, array $body = []): array
    {
        $url = $this->base . $path;
        $ch  = curl_init($url);

        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, 60);
        curl_setopt($ch, CURLOPT_HTTPHEADER, ['Content-Type: application/json']);

        $json = empty($body) ? '{}' : json_encode($body);

        if ($method === 'POST') {
            curl_setopt($ch, CURLOPT_POST, true);
            curl_setopt($ch, CURLOPT_POSTFIELDS, $json);
        } elseif ($method === 'DELETE') {
            curl_setopt($ch, CURLOPT_CUSTOMREQUEST, 'DELETE');
        }

        $raw  = curl_exec($ch);
        $code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        curl_close($ch);

        if ($raw === false) {
            throw new KrypticException('cURL request failed — is the server running?');
        }

        $data = json_decode($raw, true);
        if (!isset($data['ok']) || !$data['ok']) {
            throw new KrypticException($data['error'] ?? 'Unknown Kryptic error');
        }

        return $data;
    }

    public function get(string $path): array
    {
        $url = $this->base . $path;
        $ch  = curl_init($url);
        curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($ch, CURLOPT_TIMEOUT, 30);
        $raw = curl_exec($ch);
        curl_close($ch);
        return json_decode($raw, true) ?? [];
    }

    // ── API ──────────────────────────────────────────────────────────────────

    public function health(): array
    {
        return $this->get('/health');
    }

    public function session(): KrypticSession
    {
        $data = $this->request('POST', '/sessions');
        return new KrypticSession($this, $data['session_id']);
    }

    public function httpGet(string $url, array $headers = []): array
    {
        return $this->request('POST', '/http/get', ['url' => $url, 'headers' => $headers]);
    }

    public function httpPost(string $url, array $json = [], array $headers = []): array
    {
        return $this->request('POST', '/http/post', [
            'url'     => $url,
            'json'    => $json,
            'headers' => $headers,
        ]);
    }

    public function httpBatch(array $urls): array
    {
        $data = $this->request('POST', '/http/batch', ['urls' => $urls]);
        return $data['results'] ?? [];
    }
}

class KrypticSession
{
    private KrypticClient $client;
    public string $id;

    public function __construct(KrypticClient $client, string $id)
    {
        $this->client = $client;
        $this->id     = $id;
    }

    private function post(string $action, array $body = []): array
    {
        return $this->client->request('POST', "/sessions/{$this->id}/{$action}", $body);
    }

    private function get(string $action): array
    {
        return $this->client->get("/sessions/{$this->id}/{$action}");
    }

    public function goto(string $url, string $waitUntil = 'domcontentloaded'): array
    {
        return $this->post('goto', ['url' => $url, 'wait_until' => $waitUntil]);
    }

    public function title(): string
    {
        return $this->get('title')['title'] ?? '';
    }

    public function html(): string
    {
        return $this->get('html')['html'] ?? '';
    }

    public function url(): string
    {
        return $this->get('url')['url'] ?? '';
    }

    public function text(string $selector): string
    {
        return $this->post('text', ['selector' => $selector])['text'] ?? '';
    }

    public function click(string $selector): void
    {
        $this->post('click', ['selector' => $selector]);
    }

    public function fill(string $selector, string $value): void
    {
        $this->post('fill', ['selector' => $selector, 'value' => $value]);
    }

    public function evaluate(string $js): mixed
    {
        return $this->post('evaluate', ['js' => $js])['result'] ?? null;
    }

    public function find(string $selector): array
    {
        return $this->post('find', ['selector' => $selector]);
    }

    /**
     * Take a screenshot and return raw PNG bytes.
     */
    public function screenshot(bool $fullPage = false): string
    {
        $data = $this->post('screenshot', ['full_page' => $fullPage]);
        return base64_decode($data['data'] ?? '');
    }

    public function blockResources(array $types = ['image', 'stylesheet', 'font', 'media']): void
    {
        $this->post('block', ['resource_types' => $types]);
    }

    public function waitFor(string $selector, string $state = 'visible'): void
    {
        $this->post('wait_for', ['selector' => $selector, 'state' => $state]);
    }

    public function close(): void
    {
        $this->client->request('DELETE', "/sessions/{$this->id}");
    }
}

// ── Example ──────────────────────────────────────────────────────────────────

if (basename(__FILE__) === basename($_SERVER['SCRIPT_FILENAME'] ?? '')) {
    $k = new KrypticClient();

    $health = $k->health();
    echo "Server: " . json_encode($health) . "\n";

    $s = $k->session();
    $s->blockResources();
    $s->goto('https://example.com');
    echo "Title: " . $s->title() . "\n";
    echo "H1:    " . $s->text('h1') . "\n";
    $s->close();

    $resp = $k->httpGet('https://httpbin.org/get');
    echo "HTTP status: " . ($resp['status'] ?? '?') . "\n";

    $batch = $k->httpBatch(['https://example.com', 'https://example.org']);
    foreach ($batch as $r) {
        echo "  {$r['status']}  {$r['url']}\n";
    }
}
